include/gridtools/storage/adapter/python_sid_adapter.hpp

```diff
@@ -79,9 +85,212 @@ namespace gridtools {
                     "buffer has incorrect format: " + info.format + "; expected " + format_desc_t::format());
             return {std::move(info)};
         }
+
+        struct typestr {
```

Comment:
I hadn't realized before that this PR required the parsing of the typestr, which is kind of tedious, so kudos to your work here!

Now I'm thinking that, since the parser is already in place and `__cuda_array_interface__` and `__array_interface__` are basically equivalent, it could be possible to add another sid adaptor for CPU buffers using  `__array_interface__`. Of course this new adapter will be barely used for CPU buffers, because the Buffer protocol is faster and simpler so in practice it would be used mostly for testing by passing actual numpy arrays to the C++ side and checking that the buffers can be read and modified.

This is not really a suggestion, just a thought...